### ISSUE TITLE
Two equivalent enums are not assignable

### ISSUE THREAD

**Issue by kyeotic** [microsoft/TypeScript#49497]
# Bug Report

Two identical numeric enums cannot be assigned to each other.

### 🔎 Search Terms

* enum not assignable
* equivalent enums

### 🕗 Version & Regression Information

Current version: 4.7.2
Errors since: 3.3.3 (oldest playground version available)

### ⏯ Playground Link

[Playground with error](https://www.typescriptlang.org/play?#code/KYOwrgtgBAMglgIwCoE8AOwoG8BQUoBicATgM4AuUAvFAIwA0eUAysAMYD2IAJtVAEyN8SABYleNAMw4AvjhyhIUAIJo0qDNiZEylGgyatOPPoKajxfaXJwAbYJQBmHDgC5YiDZhrxk6YAB0OhQ4zhx8qur+QSQUQA)

[Playground without error](https://www.typescriptlang.org/play?ts=4.7.2#code/LYewJgrgNgpgBAQTgbwFBw3GAPADiAJwBcsA7CYOADRWoEYAaagJjgF9UPVRJY4AhFOkw58xMhWq0qdOAF44AchmKmVVguXNF7TqlRgYAYygBDAvCMhSAZxKnsALkQA6KgG5UV2yQBGTgTd5OAd3OAB6cLgASRIbAAsQaDA4X3hTUiwCAkIXVCA)

### 💻 Code

```
  First = 1,
  Second = 2,
  Third = 3
}

enum AppType {
  First = 1,
  Second = 2,
  Third = 3
}

let foo: LibType = LibType.First
foo = AppType.First // error: is not assignable
```

> Note: This is a minimal reproduction for a playground. In my actual case the enums have identical names and values; one is internal to a library I am using and the other is a copy in my own app.

### 🙁 Actual behavior

Error `is not assignable`

### 🙂 Expected behavior

Assignment is allowed.

For reasons I cannot identify, the expected behavior of not producing an error is [happening in this other issue](https://github.com/microsoft/TypeScript/issues/35138). This issue is actually closer to my real use case, where these two identical enums, one internal to a library and one living in my application, are not assignable to each other despite being identical in both value _and in name_. I'm curious why it fails in my actual case and in my playground, but works in [this issue](https://github.com/microsoft/TypeScript/issues/35138).

**Comment by MartinJohns** [comment]
See #33283. This is working as intended.

**Comment by fatcerberus** [comment]
Huh, someone actually found a case where numeric enums aren’t treated as synonymous with `number`… I’m impressed.

**Comment by kyeotic** [comment]
@MartinJohns 
> It works in that issue.. because it's a bug that hasn't been fixed. TypeScript accidentally believes the two X enums are the same type, when they're not.

Can you be more specific? I would like to take advantage of this incredibly useful bug, but I don't know how. What makes the bug activate?

I can assign literal values to the fields this library exports, but I can't create my own identical enum and use it in order to reduce hard-coded/magic numbers everywhere. Frankly, I thought the value of enums was in their centralization of value definitions. 

This behavior seems contrary to how every other type/value comparison happens in typescript. Two interfaces are equivalent if they have the same properties, even if they come from different libraries. Two identical type unions from different sources will be assignable two each other.  Discriminated unions from different sources will correctly narrow.

It seems that _enums alone_ are unable to assign to like-value definitions. I would really like to be able to do this, because I need to in order to use this liibrary, and if that means taking advantage of a bug I'm happy to do that.

**Comment by MartinJohns** [comment]
> What makes the bug activate?

Both enum types must have the same name and must be declared in namespaces.

> This behavior seems contrary to how every other type/value comparison happens in typescript.

Not true. Nominal typing is a thing in TypeScript (e.g. classes with non-public members). Branded-types are a concept as well (which enums are).

But yeah, `enum`s are kinda a bad feature that shouldn't have happened as-is. Violates the spirit and vision of TypeScript (as it stands for many years).

> if that means taking advantage of a bug I'm happy to do that.

Just use `// @ts-ignore`, way less painful, much more obvious. Or just create a mapping function that maps from your one type to the other (and ensures all values are mapped using the `never` type).

```typescript
function libToApp(lib: LibType): AppType {
    switch (lib) {
        case LibType.First: return AppType.First;
        case LibType.Second: return AppType.Second;
        case LibType.Third: return AppType.Third;
        default: const n: never = lib; throw new Error(`Invalid mapping.`);
    }
}
```

This is plenty of code, but **very** clear to understand and expressive.

Or if you are happy to use hacks and the names / values always match, use this: `AppType[AppType[LibType.Second] as any] as unknown as number`

**Comment by kyeotic** [comment]
I don't want to completely abandon safety with `@ts-ignore`, as that would fail to alert me if I put in the wrong enum or type entirely. The library doesn't export its enum so I can't create that mapping function, or the cast solution you provide.

This bug seems to be the only way to compile code without turning off type safety entirely.

**Comment by MartinJohns** [comment]
> This bug seems to be the only way to compile code without turning off type safety entirely.

What you're trying to do is not type-safe, so.. who gives? Just use a type assertion.

```typescript
foo = AppType.First as unknown as LibType
```

**Comment by kyeotic** [comment]
It most certainly is type safe. Assigning the number 1 to a property that accepts the number 1 is a completely valid operation. Assigning "one of 1,2,3" to a field that accepts "one of 1,2,3" is a completely valid operation. When I use a literal number `1` typescript concurs, and produces no error. It only fails when I use an enum whose value is `1` but whose providence is incorrect.

**Comment by fatcerberus** [comment]
You will find that can also assign a literal `4`, or `812`, or any other value.  Enums are not really as typesafe as they look.

**Comment by fatcerberus** [comment]
The above being said, I’m confused - you say the original enum is not exported, but yet typescript is able to typecheck a function call that involves that type.  This indicates the original enum type *is* accessible to you; you should be able to extract the type via `Parameters`.

**Comment by jcalz** [comment]
As others have mentioned, this is working as intended, for better or worse.

But if you find yourself caring about the actual literal types of the property values of an `enum` inside your TypeScript code, then it's a sign that you really don't want an `enum` at all.   Consider changing your enums to `const`-asserted objects, as mentioned [in the handbook](https://www.typescriptlang.org/docs/handbook/enums.html#objects-vs-enums):

```ts
const LibType = {
  First: 1,
  Second: 2,
  Third: 3
} as const;
type LibType = (typeof LibType)[keyof typeof LibType];

const AppType = {
  First: 1,
  Second: 2,
  Third: 3
} as const;

let foo: LibType = LibType.First
foo = AppType.First; // okay
```

[Playground link](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAMgSwEYBUCeAHApjAvDAbwCgYYAxBAJ2gC4YBGAGhJgGUtQwATOgJmdIoAFlR4wAzEQC+MAIYQYnaAG4iUTDkSoNeGAAp12EADN4ydNgCUAbQDWWNCZiGsTrRawBdVUSWwAghgYHrrEpBTUUHRMLOycYvwswqJ0kjLyiuAqREQANliwxiAgdO46+GXYAHQR0ERFILqBwRo1VCowAPSdMCC2smhAA)

**Comment by fatcerberus** [comment]
@jcalz In this case the problem is a library function that accepts an enum-typed value but doesn’t export the corresponding enum.

**Comment by kyeotic** [comment]
> That being said, this library really should have exported the enum if it intended consumers to be able to use it in API calls.

Yes, it should, its a big oversight.

@jcalz I've said it a few times, but in case you missed it, the `LibType` is not exported so I cannot use it in my own code the way your solution requires.

**Comment by 599316527** [comment]
I found that enums with same names but in different namespaces are assignable. 

Is it working as intended?

https://www.typescriptlang.org/play?ts=5.0.4#code/KYDwDg9gTgLgBAOwIYFtgGcxIMbDgZWwjGAEY4BvAKDjlEljoQFcU4A1AS3U4CNOANpxgBPSjVpwAqgDkA0jIDyAdRlwAvHAAMAGgm0AClIBCAGQCSAYQ1xSeyXAMAlc+wCCAFQCiNgEz24AF8qYKp6aHhkNEwcPEJiYF9xWnDGYBY2Lh5+IVFkyVkFFTVNXX1HEwtrTTty51dPH01-cvMZbycZN1MbAGYA4OCiBHR4JBt4klIAOiy+QWERafr3bwBuKmHRuF4ALgIiEl9Z7nncsU0kDaoAehu4AFon55fXt-eXqjDwCKZWA4SpAA+nMcot8rRCkpVDYyg4jGYrDZavCXKsmnAWrRQql4Ol-pNEiDTmC8tRIfJoSVtAEEVVkbS0Y0-HpQlsxklNITgaCFqJlkz1psICN4LxfPtCb5idk+Rc4EhfNcgA

**Comment by jcalz** [comment]
That's probably #29879
